Allow for subcanopy irrigation

fabern opened the pull request:
This PR is a draft to extend `LWFBrook90.jl` for use with subcanopy irrigation.

Subcanopy irrigation should not be added to the interception storage like normal precipitation is.
This PR is work in progress.

fabern committed:
Extend user interface to allow for irrigation

fabern commented on src/func_DiffEq_definition_cb.jl:
```diff
@@ -93,9 +93,9 @@ function LWFBrook90R_updateAmounts_INTS_INTR_SNOW_CC_SNOWLQ!(integrator)
 
     ## B) time dependent parameters
-    @unpack p_DOY, p_MONTHN, p_GLOBRAD, p_TMAX, p_TMIN, p_VAPPRES, p_WIND, p_PREC,
+    @unpack p_DOY, p_MONTHN, p_GLOBRAD, p_TMAX, p_TMIN, p_VAPPRES, p_WIND, p_PREC, p_IRRIG,
```
`p_IRRIG`, `p_δ18O_IRRIG`, `p_δ2H_IRRIG` can now be used in the callback function.

fabern commented on src/func_DiffEq_definition_f.jl:
```diff
@@ -24,9 +24,9 @@ Generate function f (right-hand-side of ODEs) needed for ODE() problem in DiffEq
         ## B) time dependent parameters
-        @unpack p_DOY, p_MONTHN, p_GLOBRAD, p_TMAX, p_TMIN, p_VAPPRES, p_WIND,  p_PREC,
+        @unpack p_DOY, p_MONTHN, p_GLOBRAD, p_TMAX, p_TMIN, p_VAPPRES, p_WIND, p_PREC, p_IRRIG,
             p_DENSEF, p_HEIGHT, p_LAI, p_SAI, p_RELDEN,
-            p_δ18O_PREC, p_δ2H_PREC, REFERENCE_DATE = p;
+            p_δ18O_PREC, p_δ2H_PREC, p_δ18O_IRRIG, p_δ2H_IRRIG, REFERENCE_DATE = p;
```
`p_IRRIG`, `p_δ18O_IRRIG`, `p_δ2H_IRRIG` can now be used in the RHS function.

fabern reviewed (commented):
See below where future work should be integrated to handle the additional input of water `p_IRRIG` with isotopic composition `p_δ18O_IRRIG`, `p_δ2H_IRRIG`. These variables are now available in the RHS and callback functions.

fabern committed:
Fix `read_path_meteoveg()`

fabern committed:
Warn when exporting an irrigation simulation to LWFBrook90R

fabern committed:
Standardize functions for csv input checking

fabern committed:
Update reference solution to make tests pass with irrigation

Equality was tested by seeing that the regression tests pass when adding. After that the reference solution was updated.

pop!(currSPAC.forcing.meteo, "p_IRRIG") # TODO: remove again
        currSPAC.forcing = NamedTuple([k=>v for (k,v) in pairs(currSPAC.forcing) if k != :irrig_iso]) # TODO: remove again
        @test loaded_currSPAC.forcing                     == currSPAC.forcing # TODO: activate again

LouisGraup committed:
Add irrigation to infiltration water and update isotopic mixing model

LouisGraup commented on src/module_ISO.jl:
```diff
@@ -210,7 +210,7 @@ end
 
 """
     compute_isotope_U_of_INTS_INTR_SNOW_and_SLFL(
```
This function seems to be unused. Implemented simple mixing model in callback function instead.

LouisGraup commented on src/func_DiffEq_definition_f.jl:
```diff
@@ -98,7 +98,7 @@ Generate function f (right-hand-side of ODEs) needed for ODE() problem in DiffEq
                     u_SWATI, p_SWATQX, p_QFPAR, p_SWATQF, p_QFFC,
                     #
-                    p_IMPERV, p_fu_RNET, aux_du_SMLT,
+                    p_IMPERV, p_IRRIG(t), p_fu_RNET, aux_du_SMLT,
```
Not sure if this is the right way to use `t` in RHS f

fabern replied:
This seems correct to me.
The function calls for RHS expected by `DifferentialEquations.jl` explicitly make reference to `t` being available. We have: `f_LWFBrook90!(du,u,p,t)`.

LouisGraup replied:
I wasn't sure if `t` would be 1 or like 1.21839 trying to reference an array with non-integer values. But it works as is, so yeah, seems fine.

fabern replied:
You are right: `t` is a `Float64`, not an integer. 
When solving the ODE, the RHS will be evaluated at arbitrary points in time, depending on the integration scheme. (Think e.g. Runge-Kutta time stepping.)

This is the reason why `p_IRRIG` is not an array but a piecewise constant interpolating function (defined [here](https://github.com/fabern/LWFBrook90.jl/blob/3804e5967981ec10ecea222e9702f264620136c5/src/func_DiffEq_definition_p.jl#L737)).
Note how Julia also distinguishes between`[` for array indexing and `(` which are function calls.

LouisGraup replied:
OK, that's interesting. I didn't realize those inputs became functions, but now that makes sense why they are referenced with `(`.

fabern commented:
Note to myself:

The CI currently fails due to loading the reference solution in [this test](https://github.com/fabern/LWFBrook90.jl/blob/a205f8b4d516231477cdf715dbdcfd5567e70c27/test/03-regression-tests.jl#L185).

I should 
- [x] check when/why this has changed 
- [x] and update the stored `*.jld2` file 
- [x] and reconsider if `*.jld2` is the best storage format. 

Note: `*.jld2` also stores the types. This had already bitten me in the past, when the SPAC type was redefined. 


```
┌ Warning: saved type DateTime has field instant::@NamedTuple{periods::@NamedTuple{value::Int64}}, but workspace type has field instant::Dates.UTInstant{Dates.Millisecond}, and no applicable convert method exists; reconstructing
└ @ JLD2 ~/.julia/packages/JLD2/wLPAN/src/data/reconstructing_datatypes.jl:213
┌ Warning: saved type NTuple{6, Symbol} has field 1::String, but workspace type has field 1::Symbol, and no applicable convert method exists; reconstructing
└ @ JLD2 ~/.julia/packages/JLD2/wLPAN/src/data/reconstructing_datatypes.jl:213
Error encountered while load FileIO.File{FileIO.DataFormat{:JLD2}, String}("../examples/DAV2020-full_u_sol_reference.jld2").

Fatal error:
regression-delta-run_example: Error During Test at /Users/runner/work/LWFBrook90.jl/LWFBrook90.jl/test/03-regression-tests.jl:185
  Got exception outside of a @test
  TypeError: in Tuple, in parameter, expected Type, got a value of type JLD2.ReconstructedMutable{Symbol("NTuple{6, Symbol}"), (Symbol("1"), Symbol("2"), Symbol("3"), Symbol("4"), Symbol("5"), Symbol("6")), NTuple{6, String}}
  Stacktrace:
```
source: https://github.com/fabern/LWFBrook90.jl/actions/runs/16969560615/job/48102545511#step:6:994

LouisGraup committed:
Ignore irrigation time-series if included as input when not simulating irrigation

LouisGraup committed:
Add irrigation to water balance calculations

LouisGraup committed:
Include irrigation in postprocessing outputs

LouisGraup committed:
Include irrigation option as output

LouisGraup committed:
Testing minor update to soil input requirements

LouisGraup committed:
Add irrigation tests

fabern committed:
fix: Avoid loading jld2 when we attempt to overwrite it

fabern committed:
test: Replace JLD2 in regression test with CSV files

fabern committed:
test: Add two simple, numeric irrigation tests (and prepare more)

fabern committed:
test: Extend number of CSV files in regression tests (include simple ones)

fabern committed:
Revert "test: Add two simple, numeric irrigation tests (and prepare more)"

This reverts commit b7f42ce8e41168ac7497f40d11354c3105f17f49.

fabern committed:
Merge pull request #97 from fabern/hotfix-JLD2

Fix CI

fabern committed:
test: Update numeric values and additional output (nothing) for IRRIG signature

fabern committed:
tests: Update Numerics of reference solution

fabern committed:
tests: small fix

fabern committed:
fix: Fix deprecation notices VA[i] => VA.u[i] and sol.destats => sol.stats

Warning: `sol.destats` is deprecated. Use `sol.stats` instead.
│   caller = ip:0x0
└ @ Core :-1

┌ Warning: `Base.getindex(VA::AbstractDiffEqArray{T, N, A}, i::Int) where {T, N, A <: Union{AbstractArray, AbstractVectorOfArray}}` is deprecated, use `VA.u[i]` instead.
│   caller = #227 at none:? [inlined]
└ @ Core ./none:-1

fabern committed:
chore: Update compat

fabern committed:
chore: Fix too restrictive compat on DiffEqBase

fabern commented on test/01-unit-tests.jl:
```diff
@@ -1019,67 +1019,72 @@ end
+    nothing nothing nothing nothing;    # TODO: do we really want to output nothing? Or simply NaN? (Prevents Union type: Union{Nothing, FLoat64})
+    nothing nothing nothing nothing;    # TODO: do we really want to output nothing? Or simply NaN? (Prevents Union type: Union{Nothing, FLoat64})
+    nothing nothing nothing nothing;    # TODO: do we really want to output nothing? Or simply NaN? (Prevents Union type: Union{Nothing, FLoat64})
+    nothing nothing nothing nothing;    # TODO: do we really want to output nothing? Or simply NaN? (Prevents Union type: Union{Nothing, FLoat64})
```
It would be ideal to define the outputs in a way that is as simple and flexible as possible for the different combinations of requested simulations:
- case 1 `simulate_isotopes = FALSE, simulate_irrigation = FALSE`
- case 2 `simulate_isotopes = TRUE, simulate_irrigation = FALSE`
- case 3 `simulate_isotopes = FALSE, simulate_irrigation = TRUE`
- case 4 `simulate_isotopes = TRUE, simulate_irrigation = TRUE`
- ... with capacitance this might add further complexity

Previously, without isotopes it did not put out these values.
Now, with isotopes but without irrigation it puts out `nothing`. 

Open questions:
1) should this be changed to NaN to keep the same type (`Float64`)? 
2) Should we have different number of outputs across cases or should we have same (with missing values as `NaN` or `missing` or `nothing`)?

It might make sense to have a test that clearly documents whether outputs are different between cases and in what way.

LouisGraup replied:
Not a bad idea to split up by cases. Fortunately, capacitance will not add further complexity in this regard as it is not a "case-by-case" basis unless we want a switch to turn it off. That also remains an open question. If cases start piling up exponentially, it would definitely be easier to just keep outputs the same and define as NaN.

The real problem with capacitance is that the integration tests will need to be redone. And the model is not backward-compatible, unless we define defaults in parameter initialization that replace missing values in param file (which would be a reasonable solution in general), or turn off capacitance by default.

fabern committed:
chore: Fix local Documenter run

fabern committed:
test: Make reference output more legible in unit tests

LouisGraup committed:
Change irrigation test for later fix with accum vars
